File: branches/cOrcs/OrcScala/src/orc/lib/state/types/CellType.scala
//
// CellType.scala -- Scala object CellType
// Project OrcScala
//
// $Id$
//
// Created by dkitchin on Dec 1, 2010.
//
// Copyright (c) 2011 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//
package orc.lib.state.types

import orc.types._
import orc.error.compiletime.typing._

/**
  *
  * @author dkitchin
  */

object CellType extends SimpleTypeConstructor("Cell", Invariant) {

  def getBuilder: Type = {
    val X = new TypeVariable()
    FunctionType(List(X), Nil, this(X))
  }

  override def instance(ts: List[Type]) = {
    val List(t) = ts
    new RecordType(
      "read" -> SimpleFunctionType(t),
      "readD" -> SimpleFunctionType(t),
      "write" -> SimpleFunctionType(t, SignalType))
  }
}

File: tags/v0.9.7/OrcJava/src/orc/lib/state/types/CellType.java
package orc.lib.state.types;

import java.util.List;

import orc.error.compiletime.typing.TypeException;
import orc.type.ArrowType;
import orc.type.DotType;
import orc.type.ListType;
import orc.type.MutableContainerType;
import orc.type.Type;

public class CellType extends MutableContainerType {

	public String toString() {
		return "Cell";
	}
	
	public Type makeCallableInstance(List<Type> params) throws TypeException {
		Type T = params.get(0);
		
		DotType dt = new DotType(/* no default behavior */);
		dt.addField("read", new ArrowType(T));
		dt.addField("readnb", new ArrowType(T));
		dt.addField("write", new ArrowType(T, Type.TOP));
		return dt;
	}
	
}
